src/unread/unreadModel.js

```diff
@@ -190,8 +190,40 @@ function streamsReducer(
       }
 
       if (action.op === 'remove') {
-        // Zulip doesn't support un-reading a message.  Ignore it.
-        return state;
+        if (action.message_info_dict === undefined) {
+          // This server is too old to support marking messages as unread.
+          return state;
+        }
+
+        let newlyUnreadState = Immutable.Map();
+
+        for (const id of action.messages) {
+          const message = action.message_info_dict?.get(id);
+
+          if (message && message.type === 'stream') {
```

Comment:
The `?.` and `message &&` indicate uncertainty about whether `action.message_info_dict` is present. But that shouldn't be necessary at this point, because we have a conditional above bailing out if it's absent -- we should be able to write this code without these extra control-flow paths to worry about.

The reason Flow isn't sure that it's present is because of that function call in between, `Immutable.Map()` -- it doesn't have a way to prove that that doesn't have side effects on the object `action`. So the usual way to fix that and make the condition be as effective as it should be is to make a local:
```js
  const { message_info_dict } = action;
  if (!message_info_dict) {
    // This server etc.
    return state;
  }
```
and use that local thereafter.

Reply:
Then I guess the `message &&` reflects another uncertainty, which is from the `get` -- it's possible that `message_info_dict` is missing something that's in `messages`.

That would be a server bug, though, IIUC. So really our logic is assuming that doesn't happen. That should be made explicit, either with a comment or by logging a warning while bailing out. Here a comment seems fine; the alternative would be more complicated, and this seems like an unlikely server bug so there's not much to gain from logging a warning for it.